File: src/interpreter.py
import operator
from .tokens import TokenType
from .nodes import (
    NumberNode,
    BinaryOpNode,
    IdentifierNode,
    AssignNode,
    BlockNode,
    ConditionalNode,
    PrintNode,
)


class Interpreter:
    def __init__(self, program):
        self.program = program
        self.symbol_table = {}

    op_map = {
        TokenType.PLUS: operator.add,
        TokenType.MINUS: operator.sub,
        TokenType.MULTIPLY: operator.mul,
        TokenType.DIVIDE: operator.truediv,
        TokenType.LESS_THAN: operator.lt,
        TokenType.GREATER_THAN: operator.gt,
        TokenType.LESS_THAN_EQUAL: operator.le,
        TokenType.GREATER_THAN_EQUAL: operator.ge,
        TokenType.EQUAL_TO: operator.eq,
        TokenType.NOT_EQUAL: operator.ne
    }

    def run(self):
        for statement in self.program.statements:
            self.evaluate(statement)


    def evaluate(self, node):
        if isinstance(node, NumberNode):
            return node.number 
        
        elif isinstance(node, IdentifierNode):
            if node.identifier in self.symbol_table:

                return self.symbol_table[node.identifier]
            else:
                raise ValueError(
                    f"ERROR: '{node.identifier}' does not exist"
                )
            
        elif isinstance(node, BinaryOpNode):
            left = self.evaluate(node.left)
            right = self.evaluate(node.right)

            return self.op_map[node.op_type.token_type](left, right)
        
        elif isinstance(node, AssignNode):
            var_name = node.var_name
            expr_eval = self.evaluate(node.expression)

            self.symbol_table[var_name] = expr_eval

        elif isinstance(node, PrintNode):
            expr_eval = self.evaluate(node.expression)
            print(f"> {expr_eval}")

        elif isinstance(node, BlockNode):
            for statement in node.statements:
                self.evaluate(statement)

        elif isinstance(node, ConditionalNode):
            result = self.evaluate(node.condition)

            if result:
                self.evaluate(node.then_block)
            else:
                if node.else_block:
                    self.evaluate(node.else_block)

        






    # def evaluate(self, node=None):
    #     if node is None:
    #         node = self.program

    #     # leaf will always be a number node so its our base case check
    #     if isinstance(node, NumberNode):
    #         return node.number

    #     left = self.evaluate(node.left)
    #     right = self.evaluate(node.right)

    #     # the call stack unwinding does the operations
    #     return self.op_map[node.op_type.token_type](left, right)

File: cpp/src/interpreter.cpp
#include "interpreter.h"
#include "nodes.h"
#include <stdexcept>
#include <iostream>
#include <string>
#include <variant>
#include <stdexcept>

/*
The Visitor Pattern:

evaluate → 
        node.accept() → 
            v.visit(*this) → 
                Interpreter::visit(ConcreteType&). 

Four hops, then the work happens.

*/

Interpreter::Interpreter() {}

const std::unordered_map<
    TokenType, std::function<std::variant<double, bool>(double, double)>> 
    Interpreter::binary_op_map = {
        { TokenType::PLUS, 
            [](double a, double b) { 
                return std::variant<double, bool>{a + b}; 
        }},
        { TokenType::MINUS, 
            [](double a, double b) { 
                return std::variant<double, bool>{a - b}; 
        }},
        { TokenType::MULTIPLY, 
            [](double a, double b) { 
                return std::variant<double, bool>{a * b}; 
        }},
        { TokenType::DIVIDE, 
            [](double a, double b) { 
                return std::variant<double, bool>{a / b}; 
        }},
        { TokenType::LESS_THAN, 
            [](double a, double b) { 
                return std::variant<double, bool>{a < b}; 
        }},
        { TokenType::GREATER_THAN, 
            [](double a, double b) { 
                return std::variant<double, bool>{a > b}; 
        }},
        { TokenType::LESS_THAN_EQUAL, 
            [](double a, double b) { 
                return std::variant<double, bool>{a <= b}; 
        }},
        { TokenType::GREATER_THAN_EQUAL, 
            [](double a, double b) { 
                return std::variant<double, bool>{a >= b}; 
        }},
        { TokenType::EQUAL_TO, 
            [](double a, double b) { 
                return std::variant<double, bool>{a == b}; 
        }},
        { TokenType::NOT_EQUAL, 
            [](double a, double b) { 
                return std::variant<double, bool>{a != b}; 
        }},
    };

const std::unordered_map<
    TokenType, std::function<std::variant<double, bool>(double)>> 
    Interpreter::unary_op_map = {
        { TokenType::MINUS, 
            [](double a) { return std::variant<double, bool>{ -a }; 
        }}
    };

// Entry point ------------------------------

void Interpreter::evaluate(ASTNode& node) {
    node.accept(*this);
}


// Evauluators ------------------------------

void Interpreter::visit(NumberNode& node) {
    result = node.number;
}

void Interpreter::visit(IdentifierNode& node) {
    auto it = symbol_table.find(node.identifier);

    if (it != symbol_table.end()) {
        result = it->second; 
    } else {
        throw std::runtime_error("ERROR: Identifier " 
            + node.identifier + " does not exist. ");
    }
}

void Interpreter::visit(BinaryOpNode& node) {
    evaluate(*node.left);
    auto left = result;

    evaluate(*node.right);
    auto right = result;

    // need .at() because map is const, need std::get for variant
    result = binary_op_map.at(node.op_type.token_type)(                                                                                                                                                         
      std::get<double>(left), std::get<double>(right)                                                                                                                                                         
    );  
}

void Interpreter::visit(UnaryOpNode& node) {
    evaluate(*node.operand);
    auto operand = result;

    result = unary_op_map.at(node.op_type.token_type)(
        std::get<double>(operand)
    );
}

void Interpreter::visit(AssignNode& node) {
    auto v_name = node.var_name;
    evaluate(*node.expression);
    auto expr = result; 

    symbol_table[v_name] = expr; 
}

void Interpreter::visit(PrintNode& node) {
    evaluate(*node.expression);
    auto expr_eval = result; 

    if (auto* d = std::get_if<double>(&expr_eval)) {    
        std::cout << "> " << *d << "\n";        
    } else if (auto* b = std::get_if<bool>(&expr_eval)) {  
        std::cout << "> " << (*b ? "True" : "False") << "\n";                                                                                                                                                                      
    }        
}

void Interpreter::visit(BlockNode& node) {
    for (const auto& stmt : node.statements) {
        evaluate(*stmt);
    }
}

void Interpreter::visit(WhileNode& node) {
    evaluate(*node.condition);
    auto cond = std::get<bool>(result);

    while (cond) {
        evaluate(*node.body_block);

         // re-evaluate the starting conditon on each iteration 
        evaluate(*node.condition);
        cond = std::get<bool>(result);
    }
}

void Interpreter::visit(ConditionalNode& node) {
    evaluate(*node.condition);
    auto cond = std::get<bool>(result);

    if (cond) {
        evaluate(*node.then_block);
    } else {
        if (node.else_block) {
            evaluate(*node.else_block);
        }
    }
}

void Interpreter::visit(ProgramNode&) {
    throw std::runtime_error("ProgramNode should not be visited directly");
}